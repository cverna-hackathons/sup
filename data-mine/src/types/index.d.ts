import { Unit } from './index';

export interface Author {
    name: string;
    foreignId?: string;
    followers?: number;
    following?: number;
    totalArtCount?: number;
    soldArtCount?: number;
}

export interface Size {
    width: number;
    height: number;
    unit: Unit;
}

export interface Entry {
    imageBase64: string;
    materials: string[];
    medias: string[];
    subjects: string[];
    styles: string[];
    size: Size;
    title?: string;
    country?: string;
    price?: number;
    author?: Author;
}

export enum QueryParam {
    material,
    media,
    subject,
    style,
}

export interface Query {
    resolutions: number[];
    materials?: string[];
    medias?: string[];
    subjects?: string[];
    styles?: string[];
}
