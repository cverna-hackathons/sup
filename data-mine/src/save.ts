import { createConnection, getRepository, ObjectType, EntityManager } from 'typeorm';
import Jimp from 'jimp';
import * as AWS from 'aws-sdk';

import { Entry, Size, Author, AWSCredentials } from './types/index.d';
import { Image } from './database/entities/Image';
import { Material } from './database/entities/Material';
import { Media } from './database/entities/Media';
import { Subject } from './database/entities/Subject';
import { Style } from './database/entities/Style';
import { Country } from './database/entities/Country';
import { Dimension } from './database/entities/Dimension';
import { Author as AuthorEntity } from './database/entities/Author';

import * as path from 'path';
import * as dotenv from 'dotenv';

const env = process.env.ENVIRONMENT || '';
const envPath = path.resolve(__dirname, `../config/${env}.env`);
dotenv.config({ path: envPath });

const WIDTH = 256;
const HEIGHT = 256;
const IMAGE_DIR = path.resolve(__dirname, `./images/`);

function getAWSCredentials(): AWSCredentials | undefined {
  const { AWS_KEY, AWS_SECRET, AWS_BUCKET } = process.env;
  if (AWS_KEY && AWS_SECRET && AWS_BUCKET) {
    return {
      key: AWS_KEY,
      secret: AWS_SECRET,
      bucket: AWS_BUCKET
    };
  }
  return undefined;
}

function awsUpload(buffer: Buffer, fileName: string, credentials: AWSCredentials): Promise<string> {
  return new Promise((resolve, reject) => {
    const s3 = new AWS.S3({
      accessKeyId: credentials.key,
      secretAccessKey: credentials.secret
    });

    s3.upload({
      Bucket: credentials.bucket,
      Key: fileName,
      Body: buffer,
    }, function(err: any, data: any) {
      if (err) {
        reject(err);
        return;
      }
      resolve(data.Location);
    });
  })
}

export async function storeImage(
  publicImageUrl: string,
  title: string = ''
): Promise<string> {
  const image = await Jimp.read(publicImageUrl);
  await image.resize(WIDTH, HEIGHT);

  const fileName = `${title}-${Date.now()}.${image.getExtension()}`;
  const credentials = getAWSCredentials();
  if (credentials) {
    const imageBuffer = await image.getBufferAsync(image.getMIME());
    return awsUpload(imageBuffer, fileName, credentials);
  } else {
    const filePath = `${IMAGE_DIR}/${fileName}`;
    await image.writeAsync(filePath);
    return filePath;
  }
}

async function getStringEntities<T>(
  entity: ObjectType<T>,
  names: string[] = []
): Promise<T[]> {
  return Promise.all(
    names.map((name: string) => {
      return getStringEntity<T>(entity, name);
    })
  );
}

async function getStringEntity<T>(
  entity: ObjectType<T>,
  name: string
): Promise<T> {
  const repo = getRepository(entity);
  const entry = await repo.findOne({
    where: [
      {
        name
      }
    ]
  });
  if (entry) {
    return Promise.resolve(entry);
  }
  const newEntry = repo.create();
  // @ts-ignore
  newEntry.name = name;
  await repo.save(newEntry);
  return newEntry;
}

async function getDimension(size: Size): Promise<Dimension> {
  const repo = getRepository(Dimension);
  const newDimension = new Dimension();
  newDimension.height = size.height;
  newDimension.width = size.width;
  newDimension.unit = size.unit;
  return repo.save(newDimension);
}

async function getAuthor(author: Author): Promise<AuthorEntity> {
  const repo = getRepository(AuthorEntity);
  const query = author.foreignId
    ? { foreignId: author.foreignId }
    : { name: author.name };
  const authorEntry = await repo.findOne({
    where: query
  });
  if (authorEntry) {
    // TODO maybe we want to update existing author with freshest data?
    return Promise.resolve(authorEntry);
  }

  const newAuthorEntry = repo.create();
  newAuthorEntry.foreignId = author.foreignId;
  newAuthorEntry.name = author.name;
  newAuthorEntry.followers = author.followers;
  newAuthorEntry.following = author.following;
  newAuthorEntry.totalArtCount = author.totalArtCount;
  newAuthorEntry.soldArtCount = author.soldArtCount;
  return repo.save(newAuthorEntry);
}

export async function saveEntries(entries: Entry[]): Promise<Image[]> {
  const { manager } = await createConnection();
  const images = await Promise.all(
    entries.map(entry => saveEntry(entry, manager))
  );

  await manager.connection.close();

  return images;
}

export async function saveEntry(entry: Entry, manager?: EntityManager): Promise<Image> {
  // TODO sanitize entry (all data lower cased, trimmed ...) so we don't have duplicates

  if (!manager) {
    manager = (await createConnection()).manager;
  }

  const image = new Image();
  image.imageUrl = await storeImage(entry.imagePublicUrl, entry.title);
  image.title = entry.title;
  image.price = entry.price;
  image.materials = await getStringEntities<Material>(
    Material,
    entry.materials
  );
  image.medias = await getStringEntities<Media>(Media, entry.medias);
  image.subjects = await getStringEntities<Subject>(Subject, entry.subjects);
  image.styles = await getStringEntities<Style>(Style, entry.styles);
  image.country = entry.country
    ? await getStringEntity<Country>(Country, entry.country)
    : undefined;
  image.dimension = await getDimension(entry.size);
  image.author = entry.author ? await getAuthor(entry.author) : undefined;

  const newImage = await manager.save(image);
  // await manager.connection.close();
  return newImage;
}
