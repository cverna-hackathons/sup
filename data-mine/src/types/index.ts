export enum Unit {
  in = 'in',
  cm = 'cm',
}
