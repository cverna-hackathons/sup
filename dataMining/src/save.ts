import { Entry } from './index.d';

export function saveEntry(entry: Entry): Promise<boolean> {

}
