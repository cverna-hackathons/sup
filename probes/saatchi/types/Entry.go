package types

type Author struct {
		Name string `json:"name"`
		ForeignId string `json:"foreignId"`
}

type Size struct {
		Width float64 `json:"width"`
		Height float64 `json:"height"`
		Unit string `json:"unit"`
}

// Entry which is saved by data-mine
type Entry struct {
    ImagePublicUrl string `json:"imagePublicUrl"`
    Materials []string `json:"materials"`
    Medias []string `json:"medias"`
    Subjects []string `json:"subjects"`
    Styles []string `json:"styles"`
    Size Size `json:"size"`
    Title string `json:"title"`
    Country string `json:"country"`
		Price float64 `json:"price"`
		Author Author `json:"author"`
}