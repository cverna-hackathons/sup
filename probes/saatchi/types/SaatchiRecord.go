package types

// Entry which is received from saatchi API
// This is transformed to Entry via transform
type SaatchiRecord struct {
		Image_url string
		Width float64
		Height float64
		Materials []string
		All_mediums []string
		All_subject string
		All_styles []string
		Artwork_title string
		Country string
		Price int
		First_name string
		Last_name string
		Id_user int
}