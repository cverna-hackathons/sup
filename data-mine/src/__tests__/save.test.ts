import { Connection, createConnection } from 'typeorm';
import * as fs from 'fs';

import { saveEntry, storeImage } from '../save';
import { Entry } from '../types/index.d';
import { Unit } from '../types/index';

describe('save', () => {
  describe('#saveEntry()', () => {
    let connection: Connection;
    beforeEach(async () => {
      connection = await createConnection();
      await connection.dropDatabase();
      await connection.synchronize();
      await connection.close();
    });
    afterAll(async () => {
      connection = await createConnection();
      await connection.dropDatabase();
      await connection.close();
    });
    it('saves image into database', async () => {
      const entry: Entry = {
        imagePublicUrl: 'pngBase64',
        materials: ['canvas'],
        medias: ['Acrylic', 'charcoal'],
        subjects: ['urban'],
        styles: ['cubism'],
        size: {
          width: 100,
          height: 150,
          unit: Unit.cm
        },
        title: 'New york in dark',
        country: 'US',
        price: 1200,
        author: {
          name: 'Harry Potter',
          foreignId: '1',
          followers: 100,
          following: 10,
          totalArtCount: 2,
          soldArtCount: 1
        }
      };

      const savedImage = await saveEntry(entry);

      expect(JSON.stringify(savedImage)).toMatchSnapshot();
    });
  });

  describe('#storeImage()', () => {
    it('stores resized image on localhost', async () => {
      const path = await storeImage(
        'https://lh3.googleusercontent.com/proxy/Y5tFZiY63lftpBgz1_k-pOGqvhvtiRr3Ou00mQfz6a3Wc5MTc7N0xgscNrgHIITbHj1cLw8UGMO4ZqFUA_DPg5Vugp3TyVatVxs',
        'Hermiona',
      );

      expect(fs.existsSync(path)).toBe(true);
    });
  });
});
