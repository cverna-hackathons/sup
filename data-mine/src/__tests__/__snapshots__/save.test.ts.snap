// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save #saveEntry() saves image into database 1`] = `"{\\"imageUrl\\":\\"DADA\\",\\"title\\":\\"New york in dark\\",\\"price\\":1200,\\"materials\\":[{\\"name\\":\\"canvas\\",\\"id\\":1}],\\"medias\\":[{\\"name\\":\\"Acrylic\\",\\"id\\":1},{\\"name\\":\\"charcoal\\",\\"id\\":2}],\\"subjects\\":[{\\"name\\":\\"urban\\",\\"id\\":1}],\\"styles\\":[{\\"name\\":\\"cubism\\",\\"id\\":1}],\\"country\\":{\\"name\\":\\"US\\",\\"id\\":1},\\"dimension\\":{\\"height\\":150,\\"width\\":100,\\"unit\\":\\"cm\\",\\"id\\":1},\\"author\\":{\\"foreignId\\":\\"1\\",\\"name\\":\\"Harry Potter\\",\\"followers\\":100,\\"following\\":10,\\"totalArtCount\\":2,\\"soldArtCount\\":1,\\"id\\":1},\\"id\\":1}"`;
